File: ios/camerawesome/Sources/camerawesome/CameraPreview/PreviewTransformPublisher.m
#import "PreviewTransformPublisher.h"

@interface PreviewTransformPublisher ()
@property(nonatomic, copy, nullable) FlutterEventSink eventSink;
@property(nonatomic, strong, nullable) NSDictionary<NSString *, id> *latestEvent;
@property(nonatomic, assign) int64_t nextSessionId;
@property(nonatomic, assign) int64_t sessionId;
@property(nonatomic, assign) int64_t revision;
@property(nonatomic, assign) int64_t textureId;
@property(nonatomic, assign) size_t bufferWidth;
@property(nonatomic, assign) size_t bufferHeight;
@property(nonatomic, assign) BOOL hasActiveSession;
@property(nonatomic, assign) BOOL hasTexture;
@property(nonatomic, assign) BOOL hasBuffer;
@property(nonatomic, assign) BOOL isMirroring;
@end

@implementation PreviewTransformPublisher

- (instancetype)init {
  self = [super init];
  if (self) {
    [[UIDevice currentDevice] beginGeneratingDeviceOrientationNotifications];
    NSNotificationCenter *notifications = [NSNotificationCenter defaultCenter];
    [notifications addObserver:self
                      selector:@selector(windowOrientationMayHaveChanged:)
                          name:UIDeviceOrientationDidChangeNotification
                        object:nil];
    [notifications addObserver:self
                      selector:@selector(windowOrientationMayHaveChanged:)
                          name:UIApplicationDidBecomeActiveNotification
                        object:nil];
    [notifications addObserver:self
                      selector:@selector(windowOrientationMayHaveChanged:)
                          name:UIWindowDidBecomeKeyNotification
                        object:nil];
  }
  return self;
}

- (void)dealloc {
  [[NSNotificationCenter defaultCenter] removeObserver:self];
  [[UIDevice currentDevice] endGeneratingDeviceOrientationNotifications];
}

- (void)setEventSink:(FlutterEventSink)eventSink {
  NSAssert([NSThread isMainThread], @"Preview transform events must use the main thread");
  _eventSink = [eventSink copy];
  if (_latestEvent != nil && _eventSink != nil) {
    _eventSink(_latestEvent);
  }
}

- (void)startSessionWithMirroring:(BOOL)isMirroring {
  NSAssert([NSThread isMainThread], @"Preview transform sessions must use the main thread");
  [self invalidateActiveSession];
  _sessionId = ++_nextSessionId;
  _revision = 0;
  _bufferWidth = 0;
  _bufferHeight = 0;
  _hasActiveSession = YES;
  _hasBuffer = NO;
  _isMirroring = isMirroring;
  [self emitInvalidated];
}

- (void)bindTextureId:(int64_t)textureId {
  NSAssert([NSThread isMainThread], @"Preview texture binding must use the main thread");
  if (!_hasActiveSession) {
    return;
  }
  if (_hasTexture && _textureId == textureId) {
    return;
  }
  _textureId = textureId;
  _hasTexture = YES;
  [self publishReadyIfPossible];
}

- (void)clearTextureBinding {
  NSAssert([NSThread isMainThread], @"Preview texture binding must use the main thread");
  _textureId = 0;
  _hasTexture = NO;
}

- (void)updateBufferWidth:(size_t)bufferWidth height:(size_t)bufferHeight {
  NSAssert([NSThread isMainThread], @"Preview buffer updates must use the main thread");
  if (!_hasActiveSession || bufferWidth == 0 || bufferHeight == 0) {
    return;
  }
  if (_hasBuffer && _bufferWidth == bufferWidth && _bufferHeight == bufferHeight) {
    return;
  }
  _bufferWidth = bufferWidth;
  _bufferHeight = bufferHeight;
  _hasBuffer = YES;
  [self publishReadyIfPossible];
}

- (void)updateMirroring:(BOOL)isMirroring {
  NSAssert([NSThread isMainThread], @"Preview mirror updates must use the main thread");
  if (!_hasActiveSession) {
    return;
  }
  if (_isMirroring == isMirroring) {
    return;
  }
  _isMirroring = isMirroring;
  [self publishReadyIfPossible];
}

- (void)invalidateActiveSession {
  NSAssert([NSThread isMainThread], @"Preview invalidation must use the main thread");
  if (!_hasActiveSession) {
    return;
  }
  _revision += 1;
  [self emitInvalidated];
  _hasActiveSession = NO;
}

- (void)windowOrientationMayHaveChanged:(NSNotification *)notification {
  dispatch_async(dispatch_get_main_queue(), ^{
    [self publishReadyIfPossible];
  });
}

- (void)publishReadyIfPossible {
  NSAssert([NSThread isMainThread], @"Preview transform publishing must use the main thread");
  if (!_hasActiveSession || !_hasTexture || !_hasBuffer) {
    return;
  }

  UIInterfaceOrientation orientation = [self currentInterfaceOrientation];
  NSInteger quarterTurns = [self quarterTurnsForOrientation:orientation];
  if (quarterTurns < 0) {
    return;
  }

  _revision += 1;
  BOOL swapsDimensions = quarterTurns % 2 == 1;
  NSDictionary<NSString *, id> *event = @{
    @"event" : @"ready",
    @"sessionId" : @(_sessionId),
    @"textureId" : @(_textureId),
    @"revision" : @(_revision),
    @"presentationQuarterTurns" : @(quarterTurns),
    @"bufferWidth" : @(_bufferWidth),
    @"bufferHeight" : @(_bufferHeight),
    @"orientedWidth" : @(swapsDimensions ? _bufferHeight : _bufferWidth),
    @"orientedHeight" : @(swapsDimensions ? _bufferWidth : _bufferHeight),
    @"cropLeft" : @0,
    @"cropTop" : @0,
    @"cropWidth" : @(_bufferWidth),
    @"cropHeight" : @(_bufferHeight),
    @"isMirroring" : @(_isMirroring),
    @"hasCameraTransform" : @YES,
  };
  [self emit:event];
}

- (UIInterfaceOrientation)currentInterfaceOrientation {
  UIApplication *application = [UIApplication sharedApplication];
  for (UIScene *scene in application.connectedScenes) {
    if (![scene isKindOfClass:[UIWindowScene class]]) {
      continue;
    }
    UIWindowScene *windowScene = (UIWindowScene *)scene;
    for (UIWindow *window in windowScene.windows) {
      if ([self controllerContainsFlutterViewController:window.rootViewController]) {
        return windowScene.interfaceOrientation;
      }
    }
  }

  id<UIApplicationDelegate> delegate = application.delegate;
  UIWindow *fallbackWindow = nil;
  if ([delegate respondsToSelector:@selector(window)]) {
    fallbackWindow = delegate.window;
  }
  if ([self controllerContainsFlutterViewController:fallbackWindow.rootViewController]) {
    return fallbackWindow.windowScene.interfaceOrientation;
  }
  return UIInterfaceOrientationUnknown;
}

- (BOOL)controllerContainsFlutterViewController:(UIViewController *)controller {
  if (controller == nil) {
    return NO;
  }
  if ([controller isKindOfClass:[FlutterViewController class]]) {
    return YES;
  }
  if ([self controllerContainsFlutterViewController:controller.presentedViewController]) {
    return YES;
  }
  for (UIViewController *child in controller.childViewControllers) {
    if ([self controllerContainsFlutterViewController:child]) {
      return YES;
    }
  }
  return NO;
}

- (NSInteger)quarterTurnsForOrientation:(UIInterfaceOrientation)orientation {
  switch (orientation) {
    case UIInterfaceOrientationPortrait:
      return 0;
    case UIInterfaceOrientationLandscapeLeft:
      return 1;
    case UIInterfaceOrientationLandscapeRight:
      return 3;
    case UIInterfaceOrientationPortraitUpsideDown:
      return 2;
    case UIInterfaceOrientationUnknown:
      return -1;
  }
}

- (void)emitInvalidated {
  [self emit:@{
    @"event" : @"invalidated",
    @"sessionId" : @(_sessionId),
    @"revision" : @(_revision),
  }];
}

- (void)emit:(NSDictionary<NSString *, id> *)event {
  _latestEvent = event;
  if (_eventSink != nil) {
    _eventSink(event);
  }
}

@end

File: android/src/main/kotlin/com/apparence/camerawesome/cameraX/preview/PreviewTransformPublisher.kt
package com.apparence.camerawesome.cameraX.preview

import android.graphics.Rect
import android.util.Log
import io.flutter.plugin.common.EventChannel

class PreviewTransformPublisher : EventChannel.StreamHandler {
    private data class Session(
        val id: Long,
        val textureId: Long,
        val bufferWidth: Int,
        val bufferHeight: Int,
        var revision: Long = 0,
    )

    private var eventSink: EventChannel.EventSink? = null
    private var latestEvent: Map<String, Any>? = null
    private var nextSessionId = 0L
    private var activeSession: Session? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        latestEvent?.let { events?.success(it) }
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    fun startSession(
        textureId: Long,
        bufferWidth: Int,
        bufferHeight: Int,
    ): Long {
        require(bufferWidth > 0 && bufferHeight > 0) {
            "Preview buffer dimensions must be positive"
        }

        invalidateActiveSession()
        val session = Session(
            id = ++nextSessionId,
            textureId = textureId,
            bufferWidth = bufferWidth,
            bufferHeight = bufferHeight,
        )
        activeSession = session
        emitInvalidated(session)
        return session.id
    }

    fun publish(
        sessionId: Long,
        transformationRotationDegrees: Int,
        sensorRotationDegrees: Int,
        cropRect: Rect,
        isMirroring: Boolean,
        hasCameraTransform: Boolean,
    ) {
        val session = activeSession
        if (session == null || session.id != sessionId) {
            return
        }

        val isFullBufferCrop = cropRect.left == 0 &&
            cropRect.top == 0 &&
            cropRect.right == session.bufferWidth &&
            cropRect.bottom == session.bufferHeight
        if (!hasCameraTransform || !isFullBufferCrop) {
            Log.w(
                TAG,
                "Rejecting preview transform session=$sessionId " +
                    "hasCameraTransform=$hasCameraTransform crop=$cropRect " +
                    "buffer=${session.bufferWidth}x${session.bufferHeight}",
            )
            session.revision += 1
            emitInvalidated(session)
            return
        }

        val quarterTurns = try {
            presentationQuarterTurns(
                transformationRotationDegrees = transformationRotationDegrees,
                sensorRotationDegrees = sensorRotationDegrees,
                isMirroring = isMirroring,
            )
        } catch (error: IllegalArgumentException) {
            Log.w(TAG, "Rejecting non-quarter preview transform", error)
            session.revision += 1
            emitInvalidated(session)
            return
        }

        session.revision += 1
        val swapsDimensions = quarterTurns % 2 == 1
        emit(
            mapOf(
                "event" to "ready",
                "sessionId" to session.id,
                "textureId" to session.textureId,
                "revision" to session.revision,
                "presentationQuarterTurns" to quarterTurns,
                "bufferWidth" to session.bufferWidth,
                "bufferHeight" to session.bufferHeight,
                "orientedWidth" to if (swapsDimensions) {
                    session.bufferHeight
                } else {
                    session.bufferWidth
                },
                "orientedHeight" to if (swapsDimensions) {
                    session.bufferWidth
                } else {
                    session.bufferHeight
                },
                "cropLeft" to cropRect.left,
                "cropTop" to cropRect.top,
                "cropWidth" to cropRect.width(),
                "cropHeight" to cropRect.height(),
                "isMirroring" to isMirroring,
                "hasCameraTransform" to hasCameraTransform,
            ),
        )
    }

    fun invalidateSession(sessionId: Long) {
        val session = activeSession
        if (session == null || session.id != sessionId) {
            return
        }
        session.revision += 1
        emitInvalidated(session)
    }

    fun invalidateActiveSession() {
        val session = activeSession ?: return
        session.revision += 1
        emitInvalidated(session)
        activeSession = null
    }

    private fun emitInvalidated(session: Session) {
        emit(
            mapOf(
                "event" to "invalidated",
                "sessionId" to session.id,
                "revision" to session.revision,
            ),
        )
    }

    private fun emit(event: Map<String, Any>) {
        latestEvent = event
        eventSink?.success(event)
    }

    private companion object {
        const val TAG = "PreviewTransform"
    }
}
